Unit tests

goodboy opened the pull request:
Like it sounds a full suite of 10 tests to make sure this bad boy works!

Appreciate any feedback you guys might have :)

Commit:
Add a sane 'sofia status' cmd wrapper

Commit:
Add a couple baseline sofia profile tests

Checks that we can push profiles from scratch and start/stop them
as well as copy from existing profiles and do the same.

Begins the long journey that is #3

Commit:
Use the proper name var for aliases

Commit:
Add a sofia profile alias test

Verifies that we can alias a profile and sofia reports it in the output
of the `sofia status` command. Note this is currently broken.
Also use the `$${local_ip_v4}` var in the profile template where
necessary.

Exposes #6

Commit:
Create sub-elements on AttrMap insertions

Commit:
Map sofia profile aliases using an AttrMap

This means we can now create profile aliases by explicitly assigning the
`name` key.

Fixes #6

Commit:
Really we're testing sofia configs

Commit:
Add a sofia gateway pushing test

Commit:
Add a list_users command parser

Add a `ConfigManager.get_users()` method which parses the output of the
`list_users` command for verifying users pushed to the "directory"
section of the configuration.

Additionally,
- add an `eval()` method to the cli object
- prefer `ConfigManager.cli` over `fscli`

Commit:
Switch the order of `KeyValues.appendfrom()` args

Commit:
Add user directory tests

These tests verify pushing users under multiple groups and that they
show up when queried using the `list_users` cli command.

Commit:
Increase default profile start/stop timeouts to 11 secs

Commit:
Add a cli factory method

Add a ``cli.get_cli()`` factory method to allow both creating new CLIs
from existing ones and to allow for creating new CLIs with different
argument inputs to ``fs_cli`` using *args and **kwargs.

Commit:
Speed up etree parsing!

Copy the remote freeswitch.xml file locally before parsing.
This drops the parse time from an avg 4 sec to about 10 ms!

Commit:
Add an event_socket.conf test

Commit:
Support sequence mappers

Some config sections are much better represented by MutableSequence
types (like sofia profile aliases). Reorg the orms class tree
to support sequences alongside mapping types whilst keeping common
etree operator logic in a base `EtreeMapper` type. Add an `ElemList`
mapper which let's you render XML elements defining a single named
attribute's value as element values.

Additionally,
- patch AttrMap to handle "emptiness" properly
- patch TagMap to handle sub-node creation correctly
- don't bother converting values to dicts passed to the __setitem__() of
  ElemMap or TagMap

Commit:
Map sofia profile aliases as a sequence

Aliases are much better managed as a mutable sequence.
Also up the default profile stopping timeout to 12 secs.

Commit:
Add more sofia tests

Add a full sofia section rendering test as well as a new aliases test
using the new sequence API.

Additionally,
- move config templates into the conftest
- use default sofia profile stop timeouts

Commit:
Add a directory section reading test

Commit:
Support private key decrypt using password prompt

Commit:
Make confmng session scoped

Commit:
Factor domain and profile fixtures into conftest.

Change `profile` to be a profile factory `mkprofile` and move the
`domain` fixture upwards for global use. Use the local IP as the domain
name.

Commit:
Add a registered user integration test

Push two profiles, one to act as a registrar with a user and the other
with a gateway that registers to that user@domain. Verify that
registration is successful using the sofia profile api.

angvp commented on sandswitches/manage.py:
```diff
@@ -63,29 +82,99 @@ def __init__(self, rfile, etree, sftp, fscli, log):
+                            argname)
+                    )
+                args.append('{} {}'.format(argname, value))
+
```
I would remove the ```if kwargs:``` condition, it's redundant, try to iterate directly .. if non extra arguments are passed then kwargs.items() is equals to []

goodboy replied:
Good call :)

vodik commented on sandswitches/orms.py:
```diff
@@ -148,17 +208,28 @@ class AttrMap(KeyValues):
-        assert len(nodes) == 1, "Not an AttrMap?"
+        nodes = self._subelems
+        if len(nodes) == 0:  # no sub-elements exists yet
+            raise KeyError
```
Is this an appropriate error to raise when there's no key?

goodboy replied:
Yeah this is for sure confusing and must have been from some kind of fix?
I can't recall atm why I did this.

goodboy replied:
Ahh yep from 13edd8f484b20396...
Yeah so looks like I can just drop `raise KeyError` stuff.

vodik commented on sandswitches/orms.py:
```diff
@@ -169,7 +240,10 @@ def __getitem__(self, key):
-        self._attribs[key] = value
+        try:
+            self._attribs[key] = value
+        except IndexError:
```
Yeah, as per comment above, we can get either `KeyError` and `IndexError` out of `__setitem__` then, basically depending on the length?

goodboy replied:
Yeah dropped the `KeyError` raise; was legacy.

vodik commented on tests/conftest.py:
```diff
@@ -0,0 +1,265 @@
+        sandswitches.comms.get_pkey(fshost, keyfile)
+        return keyfile, None
+    except Exception:
+        return keyfile, getpass.getpass("Password for {}: ".format(keyfile))
```
Will this work nicely without suspending terminal capture?

vodik replied:
Actually, wait, doesn't `get_pkey` already have away to handle the prompt for you?

And could you not catch a more specific error?

goodboy replied:
No to the first question, although I don't see why you'd try to run tests with an encrypted key if streams are being captured.

Yes `get_pkey` does do this but will prompt repeatedly due to multiple connections being created; one for each test. The idea here is that key creds are unlocked once at test session startup and then cached for further use.

Hhmm although I could probably just change the `confmng` fixture to be session scoped and maybe that willl do it. Let me try that out.

goodboy replied:
Hah yeah that worked.
Also makes the test run faster :)

vodik commented on sandswitches/comms.py:
```diff
@@ -18,7 +19,23 @@
+        if not prompt and not password:
+            raise
+        if not password and prompt:
+            password = getpass("Password for {}: ".format(keyfile))
```
Isn't this just:

```python
if not prompt:
    raise

password = getpass("Password for {}: ".format(keyfile))
return paramiko.RSAKey.from_private_key_file(keyfile, password=password)
```

goodboy replied:
Hmm yeah good point.

I was aiming for: `password` is provided and `prompt=False` -> attempt to use pw to get the key. 
Although, I guess I could just always pass `password=password` to `.from_private_key()` inside the `try:`.

vodik replied:
Ahh, didn't notice that. That said, I suspect you'll be able to pass a password to a key that isn't password locked, and it'll still work.

goodboy replied:
Yeah the default value is `None`.

goodboy commented on tests/test_register_users.py:
```diff
@@ -0,0 +1,54 @@
+
+    # register the client gateway to our 'doggy' user
+    client.register('doggy')
+    time.sleep(0.5)
```
Meant to drop this.

Commit:
Drop legacy `raise KeyError` in `AttrMap._attribs`

Not sure if we need it any more but keep this a separate commit in case
it needs to be reverted.

goodboy commented:
@vodik fixed everything I think.
Let me know :)

vodik reviewed (approved):
Looks good, nice work.
